daemon/cmd/kubeproxy_replacement_test.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	. "gopkg.in/check.v1"
+)
+
+type KubeProxySuite struct {
+	currentNetNS                  netns.NsHandle
+	testNetNS                     netns.NsHandle
+	prevConfigDevices             []string
+	prevConfigDirectRoutingDevice string
+	prevConfigEnableIPv4          bool
+	prevConfigEnableIPv6          bool
+	prevK8sNodeIP                 net.IP
+}
+
+var _ = Suite(&KubeProxySuite{})
+
+func (s *KubeProxySuite) SetUpSuite(c *C) {
+	var err error
+
+	s.prevConfigDevices = option.Config.Devices
+	s.prevConfigDirectRoutingDevice = option.Config.DirectRoutingDevice
+	s.prevConfigEnableIPv4 = option.Config.EnableIPv4
+	s.prevConfigEnableIPv6 = option.Config.EnableIPv6
+	s.prevK8sNodeIP = node.GetK8sNodeIP()
+
+	s.currentNetNS, err = netns.Get()
+	c.Assert(err, IsNil)
+	s.testNetNS, err = netns.New()
+	c.Assert(err, IsNil)
+}
+
+func (s *KubeProxySuite) TearDownTest(c *C) {
+	option.Config.Devices = s.prevConfigDevices
+	option.Config.DirectRoutingDevice = s.prevConfigDirectRoutingDevice
+	option.Config.EnableIPv4 = s.prevConfigEnableIPv4
+	option.Config.EnableIPv6 = s.prevConfigEnableIPv6
+	node.SetK8sNodeIP(s.prevK8sNodeIP)
+
+	c.Assert(s.testNetNS.Close(), IsNil)
+}
+
+func (s *KubeProxySuite) TestDetectDevices(c *C) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	c.Assert(netns.Set(s.testNetNS), IsNil)
+	defer func() { c.Assert(netns.Set(s.currentNetNS), IsNil) }()
+
+	// 1. No devices = impossible to detect
+	c.Assert(detectDevices(true, true), NotNil)
+
+	// 2. No devices, but no detection is required
+	c.Assert(detectDevices(false, false), IsNil)
+
+	// 3. Direct routing mode, should find dummy0 for both opts
+	c.Assert(createDummy("dummy0", "192.168.0.1/24"), IsNil)
+	c.Assert(createDummy("dummy1", "192.168.1.2/24"), IsNil)
+	c.Assert(createDummy("dummy2", "192.168.2.3/24"), IsNil)
+	option.Config.EnableIPv4 = true
+	option.Config.EnableIPv6 = false
+	option.Config.Tunnel = option.TunnelDisabled
+	node.SetK8sNodeIP(net.ParseIP("192.168.0.1"))
+	c.Assert(detectDevices(true, true), IsNil)
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
+
+	// 4. dummy1 should be detected too
+	c.Assert(addDefaultRoute("dummy1", "192.168.1.1"), IsNil)
+	c.Assert(detectDevices(true, true), IsNil)
+	sort.Strings(option.Config.Devices)
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0", "dummy1"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
+
+	// 5. Enable IPv6, dummy1 should not be detected, as no default route for
+	// ipv6 is found
+	option.Config.EnableIPv6 = true
+	c.Assert(detectDevices(true, true), IsNil)
+	c.Assert(option.Config.Devices, checker.DeepEquals, []string{"dummy0"})
+	c.Assert(option.Config.DirectRoutingDevice, Equals, "dummy0")
```

Comment:
I don't fully follow case 5 here. I presume the state from case 4 (ie ipv4 default route) should remain, so now that we don't have a ipv6 default route we don't care about the device with the ipv4 default route? IPv4 and IPv6 must both have default routes and both route over the same device?

Reply:
The test case does not set neither IPv6 addr nor a default IPv6 route. Anyway, due to `option.Config.EnableIPv6`,
`linuxdatapath.NodeDeviceWithDefaultRoute()` will hit the case https://github.com/cilium/cilium/blob/pr/brb/auto-multi-dev-v3/pkg/datapath/linux/route/route_linux.go#L566 which will make the former function to skip the default route device detection.